src/main/java/mate/academy/security/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,39 @@
+package mate.academy.security;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDbOptional = userService.findByEmail(email);
+        if (userFromDbOptional.isPresent()) {
+            User user = userFromDbOptional.get();
+            String hashedPassword = HashUtil.hashPassword(password, user.getSalt());
+            if (user.getPassword().equals(hashedPassword)) {
+                return user;
+            }
+        }
+        throw new AuthenticationException("Can't authenticate user with email " + email);
+    }
+
+    @Override
+    public User register(String email, String password) throws RegistrationException {
+        Optional<User> userFromDb = userService.findByEmail(email);
+        if (userFromDb.isEmpty()) {
+            User user = new User(email, password);
+            return userService.add(user);
+        }
+        throw new RegistrationException("email " + email + " was registered");
```

Comment:
Let's make this a more informations.

```suggestion
        throw new RegistrationException("User with this name " + email + "already exists.");
```